Graphics/Core/include/NovelRT/Graphics/GraphicsResource.hpp

```diff
@@ -0,0 +1,145 @@
+#pragma once
+
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
+// for more information.
+
+#include <NovelRT/Graphics/GraphicsDeviceObject.hpp>
+#include <NovelRT/Graphics/GraphicsResourceAccess.hpp>
+#include <NovelRT/Utilities/Misc.h>
+#include <cstdint>
+#include <memory>
+#include <NovelRT/Graphics/GraphicsResourceMemoryRegion.hpp>
+
+namespace NovelRT::Graphics
+{
+    template<typename TBackend> class GraphicsMemoryAllocator;
+
+    template<template <typename TBackend> typename TResource, typename TBackend> class GraphicsResourceMemoryRegion;
+
+    template<typename TBackend> struct GraphicsBackendTraits;
+
+    template<typename TBackend> class GraphicsResource : public GraphicsDeviceObject<TBackend>
+    {
+    public:
+        using BackendResourceType = typename GraphicsBackendTraits<TBackend>::ResourceType;
+
+    protected:
+        std::shared_ptr<BackendResourceType> _implementation;
+
+    private:
+        std::shared_ptr<GraphicsMemoryAllocator<TBackend>> _allocator;
+        GraphicsResourceAccess _cpuAccess;
+
+    public:
+        std::shared_ptr<GraphicsResource<TBackend>> shared_from_this()
+        {
+            return std::static_pointer_cast<GraphicsResource<TBackend>>(GraphicsDeviceObject<TBackend>::shared_from_this());
+        }
+
+        explicit GraphicsResource(std::shared_ptr<BackendResourceType> implementation,
+                                  std::shared_ptr<GraphicsMemoryAllocator<TBackend>> allocator,
+                                  GraphicsResourceAccess cpuAccess) noexcept
+            : GraphicsDeviceObject<TBackend>(allocator->GetDevice()),
+              _implementation(implementation),
+              _allocator(allocator),
+              _cpuAccess(cpuAccess)
+        {
+        }
+
+        virtual ~GraphicsResource() noexcept override = default;
+
+        //[[nodiscard]] virtual size_t GetAlignment() const noexcept = 0; //TODO: Do we still need this?
```

Comment:
Nope, probably don't need this anymore do we?

Reply:
It's not come up so most likely not.